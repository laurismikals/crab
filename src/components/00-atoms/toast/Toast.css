.toast-title {
  font-weight: 700;
}
.toast-message {
  word-wrap: break-word;
}
.toast-message a, .toast-message label {
  color: #ffffff;
}
.toast-message a:hover {
  color: #cccccc;
  text-decoration: none;
}
.toast-close-button {
  position: relative;
  right: -0.3em;
  top: -0.3em;
  float: right;
  font-size: 20px;
  font-weight: 700;
  color: #ffffff;
  text-shadow: 0 1px 0 #ffffff;
  opacity: 0.8;
  padding: 0;
  background: 0 0;
  border: 0;
  -webkit-appearance: none;
}
.toast-close-button:focus,
.toast-close-button:hover {
  color: #000000;
  text-decoration: none;
  cursor: pointer;
  opacity: 0.4;
}
.toast-top-center {
  top: 0;
  right: 0;
  width: 100%;
}
.toast-bottom-center {
  bottom: 0;
  right: 0;
  width: 100%;
}
.toast-top-full-width {
  top: 0;
  right: 0;
  width: 100%;
}
.toast-bottom-full-width {
  bottom: 0;
  right: 0;
  width: 100%;
}
.toast-top-left {
  top: 12px;
  left: 12px;
}
.toast-top-right {
  top: 12px;
  right: 12px;
}
.toast-bottom-right {
  right: 12px;
  bottom: 12px;
}
.toast-bottom-left {
  bottom: 12px;
  left: 12px;
}
.toast-container {
  position: fixed;
  z-index: 999999;
}
.toast-container * {
  box-sizing: border-box
}
.toast-container > div {
  position: relative;
  overflow: hidden;
  margin: 0 0 6px;
  padding: 15px 15px 15px 50px;
  width: 300px;
  border-radius: 3px;
  background-position: 15px center;
  background-repeat: no-repeat;
  box-shadow: 0 0 12px #999999;
  color: #ffffff;
  opacity: 0.8;
  transition: opacity 0.3s;
}
.toast--fade-out{
  opacity: 0 !important;
}
.toast-container > :hover {
  box-shadow: 0 0 12px #000000;
  opacity: 1;
  cursor: pointer;
}
.toast-container.toast-bottom-center > div, .toast-container.toast-top-center > div {
  width: 300px;
  margin: auto;
}
.toast-container.toast-bottom-full-width > div, .toast-container.toast-top-full-width > div {
  width: 96%;
  margin: auto;
}
.toast {
  background-color: #030303;
}
.toast-success {
  background-color: green;
}
.toast-error {
  background-color: red;
}
.toast-info {
  background-color: #2f96b4;
}
.toast-warning {
  background-color: orange;
}
.toast-progress {
  position: absolute;
  left: 0;
  bottom: 0;
  height: 4px;
  background-color: #000000;
  opacity: 0.4;
  width: 100%;
}
@media all and (max-width: 240px) {
  .toast-container > div {
    padding: 8px 8px 8px 50px;
    width: 11em
  }
  .toast-container .toast-close-button {
    right: -0.2em;
    top: -0.2em;
  }
}
@media all and (min-width: 241px) and (max-width: 480px) {
  .toast-container > div {
    padding: 8px 8px 8px 50px;
    width: 18em;
  }
  .toast-container .toast-close-button {
    right: -0.2em;
    top: -0.2em;
  }
}
@media all and (min-width: 481px) and (max-width: 768px) {
  .toast-container > div {
    padding: 15px 15px 15px 50px;
    width: 25em;
  }
}
