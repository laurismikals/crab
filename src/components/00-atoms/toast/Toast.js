import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import includes from 'lodash/fp/includes'
import noop from 'lodash/fp/noop'

import Interval from '../../../helpers/interval';

export class Toast extends PureComponent {
  static propTypes = {
    message: PropTypes.string.isRequired,
    type: (props, propName, componentName) => {
      if (!includes(props[propName])(['success', 'warning', 'error'])) {
        return new Error(
          'Invalid prop `' + propName + '` supplied to' +
          ' `' + componentName + '`. Validation failed.' +
          'It should be one of the following: success, warning, error.'
        );
      }
    },
    id: PropTypes.string.isRequired,
    onEnd: PropTypes.func,
    time: PropTypes.number,
  };
  static defaultProps = {
    onEnd: noop,
    time: 5000,
  };
  constructor(props){
    super(props);
    this.state = {
      initialTime: this.props.time,
      time: this.props.time,
      finished: false,
    };
  }
  removeToast() {
    this.interval.clear();

    this.setState({
      finished: true,
    });
    setTimeout(() => {
      this.props.onEnd(this.props.id);
    }, 300);
  }
  timer = () => {
    if (this.state.time <= 0) {
      this.removeToast();
    } else {
      this.setState({
        time: this.state.time - 10,
      })
    }
  };
  onClickHandler = () => {
    this.removeToast();
  };
  mouseEnterHandler = () => {
    this.interval.pause();
  };
  mouseLeaveHandler = () =>{
    this.interval.resume();
  };
  componentDidMount() {
    this.interval = new Interval(this.timer, 10);
  }
  render() {
    const { message, type } = this.props;
    const { time, finished, initialTime } = this.state;

    return (
      <div
        className={`toast toast-${type} ${finished ? 'toast--fade-out' : ''}`}
        onMouseEnter={this.mouseEnterHandler}
        onMouseLeave={this.mouseLeaveHandler}
      >
        <div
          className="toast-progress"
          style={{ width: `${time * 100 / initialTime}%`, }}
        />
        <button
          type="button"
          className="toast-close-button"
          onClick={this.onClickHandler}
        >
          Ã—
        </button>
        <div className="toast-message">
          {message}
        </div>
      </div>
    )
  }
}

export default Toast;
