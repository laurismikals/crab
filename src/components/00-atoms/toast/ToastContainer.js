import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import map from 'lodash/fp/map';

import { clearToast } from '../../../modules/toast';

import Toast from './Toast';

import './Toast.css';

export class ToastContainer extends PureComponent {
  static propTypes = {
    toasts: PropTypes.arrayOf(PropTypes.shape),
  };
  static defaultProps = {
    toasts: [],
  };
  onEndHandler(id) {
    this.props.clearToast(id);
  }
  render() {
    const { toasts } = this.props;

    return (
      <div
        id="toast-container"
        className="toast-container toast-top-right"
        aria-live="polite"
        role="alert"
      >
        {map(({ id, message, type })=> (
          <Toast
            key={id}
            id={id}
            message={message}
            type={type}
            onEnd={this.onEndHandler.bind(this)}
          />
        ))(toasts)}
      </div>
    )
  }
}

const mapStateToProps = ({
  toast: { toasts },
}) => ({
  toasts,
});

const mapDispatchToProps = {
  clearToast,
};

export default connect(mapStateToProps, mapDispatchToProps)(ToastContainer);
