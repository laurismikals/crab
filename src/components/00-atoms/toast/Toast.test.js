import React from 'react';
import Enzyme, { shallow } from 'enzyme';

import { Toast } from './Toast';
import { ToastContainer } from './ToastContainer';

describe('Atom: Toast', () => {
  const minProps = {
    message: 'This is toast message.',
    type: 'success',
    id: '1',
  };

  it('renders correctly', () => {
    const tree = shallow(
      <Toast {...minProps} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('adds class toast--fade-out when state is set to finished: true', () => {
    const tree = shallow(
      <Toast {...minProps} />
    );

    tree.setState({finished: true});

    expect(
      tree.find('.toast').hasClass('toast--fade-out')
    ).toEqual(true);
  });

});

describe('Atom: ToastContainer', () => {
  const minProps = {
    toasts: [
      {
        id: '1',
        message: 'This is toast success message.',
        type: 'success',
      },
      {
        id: '2',
        message: 'This is toast warning message.',
        type: 'warning',
      },
      {
        id: '3',
        message: 'This is toast error message.',
        type: 'error',
      }
    ],
  };

  it('renders correctly', () => {
    const tree = shallow(
      <ToastContainer {...minProps} />
    );
    expect(tree).toMatchSnapshot();
  });

});
