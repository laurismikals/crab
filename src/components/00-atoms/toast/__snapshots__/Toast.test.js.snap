// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atom: Toast renders correctly 1`] = `
<div
  className="toast toast-success "
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="toast-progress"
    style={
      Object {
        "width": "100%",
      }
    }
  />
  <button
    className="toast-close-button"
    onClick={[Function]}
    type="button"
  >
    Ã—
  </button>
  <div
    className="toast-message"
  >
    This is toast message.
  </div>
</div>
`;

exports[`Atom: ToastContainer renders correctly 1`] = `
<div
  aria-live="polite"
  className="toast-container toast-top-right"
  id="toast-container"
  role="alert"
>
  <Toast
    id="1"
    key="1"
    message="This is toast success message."
    onEnd={[Function]}
    time={5000}
    type="success"
  />
  <Toast
    id="2"
    key="2"
    message="This is toast warning message."
    onEnd={[Function]}
    time={5000}
    type="warning"
  />
  <Toast
    id="3"
    key="3"
    message="This is toast error message."
    onEnd={[Function]}
    time={5000}
    type="error"
  />
</div>
`;
